metric_tank/defcache.go

```diff
@@ -0,0 +1,134 @@
+package main
+
+import (
+	"errors"
+	"github.com/grafana/grafana/pkg/log"
+	"github.com/raintank/raintank-metric/metricdef"
+	"github.com/raintank/raintank-metric/schema"
+	"sync"
+	"time"
+)
+
+// design notes:
+// MT pulls in all definitions when it starts up.
+// those "old" ones + whatever it sees as inputs from the metrics queue
+// is enough for it to always know the complete current state
+// nothing should update ES "behind its back", so we never need to pull
+// from ES other then at startup.
+// but other MT instances may update ES while we are down, so ES is a good
+// place to pull from, until the performance is demonstrably too slow.
+// there are some vectors here for race conditions but we can work those out
+// later, perhaps when tacking the multiple-intervals work
+
+type DefCache struct {
+	sync.RWMutex
+	defs map[string]*schema.MetricDefinition
+}
+
+func NewDefCache() *DefCache {
+	d := &DefCache{
+		defs: make(map[string]*schema.MetricDefinition),
+	}
+	go d.Backfill()
+	return d
+}
+
+// backfill definitions from ES
+// in theory, there is a race between defs from ES and from nsq
+// in practice, it doesn't matter: you're only supposed to query MT
+// after a while, after which the defs surely have stabilized.
+func (dc *DefCache) Backfill() {
+	total := 0
+	add := func(met []*schema.MetricDefinition) {
+		if len(met) > 0 {
+			total += len(met)
+			dc.Lock()
+			for _, def := range met {
+				dc.defs[def.Id] = def
+			}
+			dc.Unlock()
+		}
+	}
+	met, scroll_id, err := metricdef.GetMetrics("")
+	if err != nil {
+		log.Error(3, "Could not backfill from ES: %s", err)
+		return
+	}
+	add(met)
+	for scroll_id != "" {
+		met, scroll_id, err = metricdef.GetMetrics(scroll_id)
+		if err != nil {
+			log.Error(3, "Could not backfill from ES: %s", err)
+			return
+		}
+		add(met)
+	}
+	log.Debug("backfilled %d metric definitions", total)
+}
+
+func (dc *DefCache) Add(metric *schema.MetricData) {
+	id := metric.GetId()
+	dc.Lock()
+	mdef, ok := dc.defs[id]
+	dc.Unlock()
+	if ok {
+		if mdef.LastUpdate < time.Now().Unix()-600 {
```

Comment:
after reflection, i think that we need to keep updating the "lastUpdate" field so that we can identify stale metrics and not include them in render requests.  Ie, if a metric has not been written to for a week we should not be returning it when rendering the last 1day.

But, i think we need to increase the update interval from 10mins to maybe 6 or 12hours.

I think this can all be significantly improved with some re-factoring, but i would prefer to delay doing that until after we have moved to Kafka